### ISSUE TITLE
`ModelCheckpoint` not working when giving unexpected arg to `Trainer`

### ISSUE THREAD

**Issue by omerferhatt** [Lightning-AI/lightning#15180]
### Bug description

Model checkpoint not working when `val_check_interval` arg set while `validation_dataloader` given empty. But `val_check_interval` removed from args it works normally. They shouldn't block each other, since checkpoint callback works according to given metric name.

### How to reproduce the bug

```python
callbacks = [ModelCheckpoint(
    monitor="Some/Train/Metric",
    mode="max",
    verbose=True,
    save_last=True,
    save_weights_only=False,
)]

datamodule # Train only datamodule

trainer = pl.Trainer(
    ...
    val_check_interval=0.5,
    ...
)  # Broken

trainer = pl.Trainer(
    ...
    # val_check_interval=0.5,
    ...
)  # Works normal
```

### Error messages and logs

There's no error message.

### Environment

```
* CUDA:
        - GPU:
                - NVIDIA A100-SXM4-40GB
                - NVIDIA A100-SXM4-40GB
        - available:         True
        - version:           11.6
* Lightning:
        - efficientnet-pytorch: 0.7.1
        - pytorch-lightning: 1.7.7
        - pytorch-ranger:    0.1.1
        - segmentation-models-pytorch: 0.3.0
        - torch:             1.12.1+cu116
        - torch-optimizer:   0.3.0
        - torchaudio:        0.12.1+cu116
        - torchmetrics:      0.9.3
        - torchvision:       0.13.1+cu116
* Packages:
        - absl-py:           1.3.0
        - aiohttp:           3.8.3
        - aiosignal:         1.2.0
        - albumentations:    1.3.0
        - alpha-trainer:     2.4.0
        - antlr4-python3-runtime: 4.9.3
        - async-timeout:     4.0.2
        - attrs:             22.1.0
        - cachetools:        5.2.0
        - certifi:           2022.9.24
        - charset-normalizer: 2.1.1
        - checksumdir:       1.2.0
        - click:             8.1.3
        - coloredlogs:       15.0.1
        - colorlog:          6.7.0
        - commonmark:        0.9.1
        - docker-pycreds:    0.4.0
        - efficientnet-pytorch: 0.7.1
        - filelock:          3.8.0
        - flatbuffers:       22.9.24
        - frozenlist:        1.3.1
        - fsspec:            2022.8.2
        - gitdb:             4.0.9
        - gitpython:         3.1.29
        - google-api-core:   2.10.2
        - google-auth:       2.13.0
        - google-auth-oauthlib: 0.4.6
        - google-cloud-core: 2.3.2
        - google-cloud-storage: 2.5.0
        - google-crc32c:     1.5.0
        - google-resumable-media: 2.4.0
        - googleapis-common-protos: 1.56.4
        - grpcio:            1.50.0
        - huggingface-hub:   0.10.1
        - humanfriendly:     10.0
        - hydra-colorlog:    1.2.0
        - hydra-core:        1.2.0
        - idna:              3.4
        - imageio:           2.22.2
        - importlib-metadata: 5.0.0
        - importlib-resources: 5.10.0
        - iterative-stratification: 0.1.7
        - joblib:            1.2.0
        - kornia:            0.6.8
        - markdown:          3.4.1
        - markupsafe:        2.1.1
        - mpmath:            1.2.1
        - multidict:         6.0.2
        - munch:             2.5.0
        - networkx:          2.8.7
        - numpy:             1.23.4
        - oauthlib:          3.2.2
        - omegaconf:         2.2.3
        - onnx:              1.12.0
        - onnxruntime:       1.12.1
        - onnxsim:           0.4.8
        - opencv-python-headless: 4.6.0.66
        - packaging:         21.3
        - pandas:            1.5.0
        - pathtools:         0.1.2
        - pillow:            9.2.0
        - pip:               22.3
        - pkg-resources:     0.0.0
        - pretrainedmodels:  0.7.4
        - promise:           2.3
        - protobuf:          3.19.6
        - psutil:            5.9.3
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.8
        - pydeprecate:       0.3.2
        - pygments:          2.13.0
        - pyparsing:         3.0.9
        - python-dateutil:   2.8.2
        - pytorch-lightning: 1.7.7
        - pytorch-ranger:    0.1.1
        - pytz:              2022.5
        - pywavelets:        1.4.1
        - pyyaml:            6.0
        - qudida:            0.0.4
        - regex:             2022.9.13
        - requests:          2.28.1
        - requests-oauthlib: 1.3.1
        - rich:              12.6.0
        - rsa:               4.9
        - scikit-image:      0.19.3
        - scikit-learn:      1.1.2
        - scipy:             1.9.2
        - segmentation-models-pytorch: 0.3.0
        - sentry-sdk:        1.9.10
        - setproctitle:      1.3.2
        - setuptools:        65.5.0
        - shortuuid:         1.0.9
        - six:               1.16.0
        - smmap:             5.0.0
        - sympy:             1.11.1
        - tensorboard:       2.10.1
        - tensorboard-data-server: 0.6.1
        - tensorboard-plugin-wit: 1.8.1
        - threadpoolctl:     3.1.0
        - tifffile:          2022.10.10
        - timm:              0.4.12
        - tokenizers:        0.13.1
        - torch:             1.12.1+cu116
        - torch-optimizer:   0.3.0
        - torchaudio:        0.12.1+cu116
        - torchmetrics:      0.9.3
        - torchvision:       0.13.1+cu116
        - tqdm:              4.64.1
        - transformers:      4.23.1
        - typing-extensions: 4.4.0
        - urllib3:           1.26.12
        - wandb:             0.13.4
        - werkzeug:          2.2.2
        - wheel:             0.37.1
        - yarl:              1.8.1
        - zipp:              3.9.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #24~20.04.1-Ubuntu SMP Mon Sep 12 06:14:01 UTC 2022
```

**Comment by awaelchli** [comment]
This issue stems from these lines here:

https://github.com/Lightning-AI/lightning/blob/2d3db64b7c00bb6b5f8dcca395a46cc7368df665/src/pytorch_lightning/callbacks/model_checkpoint.py#L267-L270

The argument ModelCheckpoint(save_on_train_epoch_end=True|False) determines whether it should save on the training epoch level OR save according to the validation intervals. If the users doesn't set it, the value is determined automagically. 

Given @omerferhatt's observations, this does seem weird at first. I'm not sure how we can work around this correctly. 
On the other hand, I'm wondering whether the setting `Trainer(val_check_interval=x)` AND no validation loader/step should even be a valid configuration? Perhaps this shouldn't be allowed in the first place, as we are requesting to validate when it is not possible.

cc @rohitgr7 @carmocca who might be interested in such discussion.

**Comment by rohitgr7** [comment]
this might conflict with `reload_dataloaders_every_n_epochs` where user have an option to start validation after certain epochs..

how about we make a method `should_save_on_train_epoch_end `? and there we can check both `val_check_interval` and `num_val_batches` every time?

**Comment by rohitgr7** [comment]
@omerferhatt are you interested in sending a fix for it? :)

**Comment by awaelchli** [comment]
> how about we make a method should_save_on_train_epoch_end ? and there we can check both val_check_interval and num_val_batches every time?

If this is self-contained within the Callback, I'm ok with it.
